local.py

```diff
@@ -0,0 +1,86 @@
+# Before executing lines of this script at a time, run `/start_postrgres.sh`.
+
+import os
+
+import sqlalchemy as sa
+
+# We will first connect to database with the default user.
+# Once connected, we will create two tables and a user 'jim'.
+# jim should have all rights for table 'companies' but no
+# rights for table 'companies_secret'.
+
+# Step 1: Set up the database.
+
+address = os.environ.get("DB_ADDR", "localhost")
+connection_string = f"postgresql://datajudge:datajudge@{address}:5432/datajudge"
+engine = sa.create_engine(connection_string)
+
+with engine.connect() as con:
+    con.execute(
+        "CREATE TABLE companies (id INTEGER PRIMARY KEY, name TEXT, num_employees INTEGER)"
+    )
+    con.execute(
+        "INSERT INTO companies (id, name, num_employees) VALUES (1, 'QuantCo', 100), (2, 'Google', 150000), (3, 'BMW', 120000), (4, 'Apple', 145000)"
+    )
+
+    con.execute(
+        "CREATE TABLE companies_secret (id INTEGER PRIMARY KEY, name TEXT, num_employees INTEGER)"
+    )
+    con.execute(
+        "INSERT INTO companies_secret (id, name, num_employees) VALUES (1, 'QuantCo', 100), (2, 'Google', 150000), (3, 'BMW', 120000), (4, 'Apple', 145000)"
+    )
+
+    con.execute("CREATE ROLE jim  LOGIN PASSWORD 'jim'")
+    con.execute("GRANT ALL on companies TO jim")
+    con.execute("REVOKE ALL ON companies_secret FROM jim")
+
+
+# Step 2: Query the database.
+
+connection_string_jim = f"postgresql://jim:jim@{address}:5432/datajudge"
+engine_jim = sa.create_engine(connection_string_jim)
+
+with engine_jim.connect() as con:
+    result1 = con.execute("SELECT * FROM companies").fetchall()
+
+print(result1)
+
+
+# This fails.
+with engine_jim.connect() as con:
+    result2 = con.execute("SELECT * FROM companies_secret").fetchall()
+
+print(result2)
+
+metadata = sa.MetaData(engine_jim)
+
+# Both of these succeed. Even though jim doesn't have rights to read from
+# 'companies_secret', all of its metadata around columns, constraints etc.
+# can be investigaed.
+table = sa.Table("companies", metadata, autoload_with=engine_jim)
+table_secret = sa.Table("companies_secret", metadata, autoload_with=engine_jim)
+
+with engine_jim.connect() as con:
+    try:
+        result2 = con.execute("SELECT * FROM companies").first()
+    except sa.exc.ProgrammingError as error:
+        if "psycopg2.errors.InsufficientPrivilege" in str(error):
+            print("missing privilege")
+        else:
+            raise error
+
+
+with engine.connect() as con:
+    result = con.execute(
+        "SELECT table_name, grantee, privilege_type FROM information_schema.table_privileges WHERE grantee = 'jim'"
```

Comment:
@svengiegerich 
A quick google search makes it look as though this schema and table were available across many dialects/dbms.
Afaik sqlalchemy relies on it to do inspection.

Reply:
> USE my_db
GO
SELECT *
FROM INFORMATION_SCHEMA.TABLE_PRIVILEGES

Note that you are using `my_db` when looking for `FROM INFORMATION_SCHEMA.TABLE_PRIVILEGES`. I don't know about this but I could very well imagine this schema to live in `tempdb` on mssql. :)

Reply:
At first sight it does indeed look like that. See this reference in sqlalchemy:
https://github.com/sqlalchemy/sqlalchemy/blob/d305f0130dd1cf77f8ffe9f22e63dd770630ec81/lib/sqlalchemy/dialects/mssql/information_schema.py#L107

Reply:
More context on access to `INFORMATION_SCHEMA` on mssql:

> The visibility of the metadata in information schema views is limited to securables that a user either owns or on which the user has been granted some permission.

https://learn.microsoft.com/en-us/sql/relational-databases/system-information-schema-views/system-information-schema-views-transact-sql?view=sql-server-ver16#permissions